nixos/modules/services/networking/envoy.nix

```diff
@@ -0,0 +1,74 @@
+{ config, lib, pkgs, ... }:
+
+with lib;
+
+let
+  cfg = config.services.envoy;
+  format = pkgs.formats.json { };
+  conf = format.generate "envoy.json" cfg.settings;
+  validateConfig = file:
+    pkgs.runCommand "validate-envoy-conf" { } ''
+      ${pkgs.envoy}/bin/envoy --log-level error --mode validate -c "${file}"
+      cp "${file}" "$out"
+    '';
+
+in
+
+{
+  options.services.envoy = {
+    enable = mkEnableOption "Envoy reverse proxy";
+
+    settings = mkOption {
+      type = (pkgs.formats.json { }).type;
+      default = { };
+      example = literalExpression ''
+        {
+          admin = {
+            access_log_path = "/dev/null";
+            address = {
+              socket_address = {
+                protocol = "TCP";
+                address = "127.0.0.1";
+                port_value = 9901;
+              };
+            };
+          };
+          static_resources = {
+            listeners = [];
+            clusters = [];
+          };
+        }
+      '';
+      description = ''
+        Specify the configuration for Envoy in Nix.
+      '';
+    };
+  };
+
+  config = mkIf cfg.enable {
+    environment.systemPackages = [ pkgs.envoy ];
+    users.groups._envoy = { };
+    users.users._envoy = {
+      description = "Envoy service user";
+      group = "_envoy";
+      isSystemUser = true;
+    };
+    systemd.services.envoy = {
+      description = "Envoy reverse proxy";
+      after = [ "network-online.target" ];
+      requires = [ "network-online.target" ];
+      wantedBy = [ "multi-user.target" ];
+      serviceConfig = {
+        ExecStart = "${pkgs.envoy}/bin/envoy -c ${validateConfig conf}";
+        User = "_envoy";
+        Group = "_envoy";
```

Comment:
best to remove these two. if this user/group is configured on the system then they'll be used, which probably won't be the intention of the user?
```suggestion
```

systemd will attempt to allocate a temporary `envoy` user/group automatically if these are not set, adding a hash if necessary to get a unique name.

Reply:
not sure why we would to be honest, is there something i'm missing?

Reply:
most distros I have come across (e.g. Debian, Void) use an underscore prefix for users they create to avoid conflicts with users created by admins

>When maintainers choose a new hardcoded or dynamically generated username for packages to use, they should start this username with an underscore. This minimizes collisions with locally created user accounts.

https://www.debian.org/doc/debian-policy/ch-opersys.html#introduction

Reply:
if you don't set User/Group systemd gurantees that no conflict will happen. for statically allocated names that may be a sensible approach, but here it's not necessary.

nixos doesn't follow this convention (in most places).

Reply:
alright then I removed them... just seems like a lot of distros are doing this (even for dynamic users) and it does not seem like a bad idea.